README.md
# rucrdtw - R Bindings for the UCR Suite

<!-- badges: start -->

[![R-CMD-check](https://github.com/pboesu/rucrdtw/workflows/R-CMD-check/badge.svg)](https://github.com/pboesu/rucrdtw/actions)
[![codecov](https://codecov.io/gh/pboesu/rucrdtw/branch/master/graph/badge.svg)](https://app.codecov.io/gh/pboesu/rucrdtw)
[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/rucrdtw)](https://cran.r-project.org/package=rucrdtw)
[![Peer_Reviewed_Software](https://img.shields.io/badge/Peer%20Reviewed-%E2%9C%93-green.svg)](http://doi.org/10.21105/joss.00100)
[![DOI:10.21105/joss.00100](https://img.shields.io/badge/DOI-10.21105%2Fjoss.00100-blue.svg)](http://doi.org/10.21105/joss.00100)
<!-- badges: end -->

R bindings for functions from the [UCR
Suite](http://www.cs.ucr.edu/~eamonn/UCRsuite.html) (Thanawin
Rakthanmanon et al. 2012 SIGKDD), which enables ultrafast subsequence
search under both Dynamic Time Warping and Euclidean Distance. UCR DTW
is several orders of magnitudes faster than naive DTW searches.

![](inst/img/dtw-comparison-1.png)

## Installation

Install `rucrdtw` from GitHub:

``` r
install.packages("devtools")
devtools::install_github("pboesu/rucrdtw")
```

## Examples

Examples are contained in the vignette `rucrdtw`:

``` r
library("rucrdtw")
vignette("using_rucrdtw")
```

## Citation

To cite rucrdtw in publications, please use:

Boersch-Supan (2016). rucrdtw: Fast time series subsequence search in R.
The Journal of Open Source Software <http://doi.org/10.21105/joss.00100>

Rakthanmanon et al. (2012). Searching and mining trillions of time
series subsequences under dynamic time warping. SIGKDD
<http://doi.org/10.1145/2339530.2339576>

## Bug reports and contributions

Please file a github issue at <https://github.com/pboesu/rucrdtw/issues>
if you find any problems or have feature suggestions. Contributions (via
pull requests or otherwise) are welcome. Read more about how to
contribute and the code of conduct for contributions
[here](https://github.com/pboesu/rucrdtw/blob/master/CONTRIBUTING.md).
If you don’t like github you can contact the package maintainer at
<pboesu@gmail.com>.

## License

COPYRIGHT 2012: Thanawin Rakthanmanon, Bilson Campana, Abdullah Mueen,
Gustavo Batista and Eamonn Keogh. (UCR Suite source code)

COPYRIGHT 2016: Philipp Boersch-Supan (R bindings)

Licensed under the Apache License, Version 2.0 (the “License”); you may
not use this file except in compliance with the License. You may obtain
a copy of the License at

<https://www.apache.org/licenses/LICENSE-2.0>

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an “AS IS” BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.


NEWS.md
# rucrdtw 0.1.6

* updated to Rcpp 1.0.11.6 to address -Wformat-security issue

# rucrdtw 0.1.5

* updated documentation to address breaking change in roxygen2
* require R >= 3.6 to ensure C++11 standard

# rucrdtw 0.1.4

* updated code in anticipation of stringsAsFactors = FALSE in forthcoming R versions
* adjusted examples and vignettes to accommodate run-times of examples that are numerically zero
* updated citation information in README


# rucrdtw 0.1.3

* Fixed memory leaks in C++ routines

# rucrdtw 0.1.2

* Added a `NEWS.md` file to track changes to the package
* Updated CITATION to reflect publication in JOSS
* C++ routines are registered now
* vignette now uses the rbenchmark package instead of microbenchmark




tests/testthat.R
library(testthat)
library(rucrdtw)

test_check("rucrdtw")


tests/testthat/test-errors.R
context("test failures on inputs")

#get test data file locations
dataf <- system.file("extdata/col_sc.txt", package="rucrdtw")
firstf <- system.file("extdata/first_sc.txt", package="rucrdtw")
#load data
data("synthetic_control")
#get query length
qlength <- length(synthetic_control[1,])
#cut off some columns
wrong_ref <- synthetic_control[,1:55]

test_that("error message is displayed on unsuitable input", {
  #input not a numeric matrix
  expect_error(ucrdtw_mv(dataf, firstf, 0.05), "data must be a numeric matrix")
  expect_error(ucred_mv(dataf, firstf), "data must be a numeric matrix")
  expect_error(ucrdtw_ff("missing.file", firstf, 60, 0.05), "ERROR : File not Found!!!")
  expect_error(ucrdtw_ff(dataf, "missing.file", 60, 0.05), "ERROR : File not Found!!!")
  expect_error(ucred_ff("missing.file", firstf, 60), "ERROR : File not Found!!!")
  expect_error(ucred_ff(dataf, "missing.file", 60), "ERROR : File not Found!!!")
  expect_error(ucrdtw_fv("missing.file", synthetic_control[1,], 0.05), "ERROR : File not Found!!!")
  expect_error(ucred_fv("missing.file", synthetic_control[1,]), "ERROR : File not Found!!!")
})

test_that("dtw vm method errors on dimension mismatch", {
    query <- synthetic_control[123,]
    expect_error(ucrdtw_mv(wrong_ref, query, 0.05, byrow = TRUE), "data series length must match query length")
    expect_error(ucrdtw_mv(wrong_ref, query, 0.05, byrow = FALSE), "data series length must match query length")
})

test_that("dtw ed method works errors on dimension mismatch", {
  query <- synthetic_control[123,]
  expect_error(ucred_mv(wrong_ref, query, byrow = TRUE), "data series length must match query length")
  expect_error(ucred_mv(wrong_ref, query, byrow = FALSE), "data series length must match query length")
})


tests/testthat/test-summaries.R
context("test summary output")

#get test data file locations
dataf <- system.file("extdata/col_sc.txt", package="rucrdtw")
firstf <- system.file("extdata/first_sc.txt", package="rucrdtw")

test_that("ucrdtw summary method works", {
  first = ucrdtw_ff(dataf, firstf, 60, 0.05)
  x <- summary(first)
  expect_equal(class(x), "data.frame")
  expect_equal(ncol(x), length(first))
})


test_that("ucred summary method works", {
  first = ucred_ff(dataf, firstf, 60)
  x <- summary(first)
  expect_equal(class(x), "data.frame")
  expect_equal(ncol(x), length(first))
})


R/ucrdtw.R
#' Summarize subsequence search
#'
#' Summary method for class ucrdtw
#'
#' @param object an object of class ucrdtw
#' @param ... further arguments passed to or from methods
#'
#' @return An unlisted version of the object is returned.
#' @export
#'
summary.ucrdtw <- function(object, ...){
  print(as.data.frame.list(object), row.names=FALSE, stringsAsFactors = FALSE, ...)
}

#' Summarize subsequence search
#'
#' Summary method for class ucred
#'
#' @param object an object of class ucred
#' @param ... further arguments passed to or from methods
#'
#' @return An unlisted version of the object is returned.
#' @export
#'
summary.ucred <- function(object, ...){
  print(as.data.frame.list(object), row.names=FALSE, stringsAsFactors = FALSE, ...)
}


#' UCR DTW Algorithm matrix-vector method
#'
#' This implementation of the UCR Suite command line utility, takes an R numeric matrix as data input and an R numeric vector for the query.
#' The default behaviour differs from the other methods, in that it does not perform a sliding window search for a match.
#' Instead it is designed to find a best match for a query in a reference set of time-series of the same length as the query.
#' This is useful, for example, when comparing a time-series of undetermined class to a labelled reference set of classified time-series.
#'
#' @name ucrdtw_mv
#' @param data numeric matrix containing data
#' @param query numeric vector containing the query. This determines the query length.
#' @param dtwwindow double; Size of the warping window size (as a proportion of query length). The DTW calculation in `rucrdtw` uses a symmetric Sakoe-Chiba band. See Giorgino (2009) for a general coverage of warping window constraints.
#' @param epoch int defaults to 1e5, should be \eqn{\le} 1e6. This is the size of the data chunk that is processed at once. All cumulative values in the algorithm will be restarted after \code{epoch} iterations to reduce floating point errors in these values.
#' @param skip boolean; defaults to TRUE. If TRUE bound calculations and if necessary, distance calculations, are only performed on non-overlapping segments of the data (i.e. multiples of \code{length(query)}). This is useful if \code{data} is a set of multiple reference time series, each of length \code{length(query)}. The location returned when skipping is the index of the subsequence.
#' @param byrow logical; If TRUE rows in \code{data} represent time-series, columns time-points
#' @return a ucrdtw object. A list containing the following elements
#' \itemize{
#'   \item \strong{location:} The row or column number of the nearest neighbor of the given query in the data set.
#'   \item \strong{distance:} The DTW distance between the nearest neighbor and the query.
#'   \item \strong{prunedKim:} Percentage of subsequences that were pruned based on the LB-Kim criterion.
#'   \item \strong{prunedKeogh:} Percentage of subsequences that were pruned based on the LB-Keogh-EQ criterion.
#'   \item \strong{prunedKeogh2:} Percentage of subsequences that were pruned based on the LB-Keogh-EC criterion.
#'   \item \strong{dtwCalc:} Percentage of subsequences for which the full DTW distance was calculated.
#' }
#' @references Giorgino, Toni (2009). Computing and Visualizing Dynamic Time Warping Alignments in R: The dtw Package. Journal of Statistical Software, 31(7), 1-24, doi:\doi{10.18637/jss.v031.i07}.
#' @examples
#' #load example data
#' data("synthetic_control")
#' query <- synthetic_control[5,]
#' #run query
#' ucrdtw_mv(synthetic_control, query, 0.05, byrow = TRUE)
#' @export
ucrdtw_mv <- function(data, query, dtwwindow, epoch = 1e5, skip=TRUE, byrow=FALSE){
  if (!is.matrix(data) && !is.numeric(data)) stop("data must be a numeric matrix")
  if(byrow){
    if(ncol(data)!=length(query)) stop("data series length must match query length")
    datav <- as.vector(t(data))
  } else {
    if(nrow(data)!=length(query)) stop("data series length must match query length")
    datav <- as.vector(data)
  }
  ucrdtw_vv(data=datav, query, dtwwindow, epoch = 1e5, skip=TRUE)
}


#' UCR ED Algorithm matrix-vector method
#'
#' This implementation of the UCR Suite command line utility, takes an R numeric matrix as data input and an R numeric vector for the query.
#' The default behaviour differs from the other methods, in that it does not perform a sliding window search for a match.
#' Instead it is designed to find a best match for a query in a reference set of time-series of the same length as the query.
#' This is useful, for example, when comparing a time-series of undetermined class to a labelled reference set of classified time-series.
#'
#' @name ucred_mv
#' @param data numeric matrix containing data
#' @param query numeric vector containing the query. This determines the query length.
#' @param skip boolean; defaults to TRUE. If TRUE bound calculations and if necessary, distance calculations, are only performed on non-overlapping segments of the data (i.e. multiples of \code{length(query)}). This is useful if \code{data} is a set of multiple reference time series, each of length \code{length(query)}. The location returned when skipping is the index of the subsequence.
#' @param byrow logical; If TRUE rows in \code{data} represent time-series, columns time-points
#' @return a ucred object. A list with the following elements
#' \itemize{
#'   \item \strong{location:} The starting location of the nearest neighbor of the given query, of size \code{qlength}, in the data. Note that location starts from 1.
#'   \item \strong{distance:} The euclidean distance between the nearest neighbor and the query.
#' }
#' @examples
#' #load example data matrix
#' data("synthetic_control")
#' #use on arbitrary row as query
#' query <- synthetic_control[5,]
#' #run query
#' ucred_mv(synthetic_control, query, byrow=TRUE)
#' @export
ucred_mv <- function(data, query, skip=TRUE, byrow=FALSE){
  if (!is.matrix(data) && !is.numeric(data)) stop("data must be a numeric matrix")
  if(byrow){
    if(ncol(data)!=length(query)) stop("data series length must match query length")
    datav <- as.vector(t(data))
  } else {
    if(nrow(data)!=length(query)) stop("data series length must match query length")
    datav <- as.vector(data)
  }
  ucred_vv(data=datav, query, skip=TRUE)
}


R/data_doc.R
#' Synthetic Control Chart Time Series Data Set
#'
#' This dataset contains 600 examples of control charts synthetically generated by the process in Alcock and Manolopoulos (1999) and obtained from the UCI Machine Learning Repository \url{http://archive.ics.uci.edu/ml/datasets/Synthetic+Control+Chart+Time+Series}.
#'
#' There are six different classes of control charts:
#' \enumerate{
#'      \item Normal (Rows 1-100)
#'      \item Cyclic (Rows 101-200)
#'      \item Increasing trend (Rows 201-300)
#'      \item Decreasing trend (Rows 301-400)
#'      \item Upward shift (Rows 401-500)
#'      \item Downward shift (Rows 501-600)
#'      }
#'
#' @name synthetic_control
#' @docType data
#' @format A matrix with 600 rows (series) and 60 columns (timepoints)
#'
#' @references Alcock R.J. and Manolopoulos Y. Time-Series Similarity Queries Employing a Feature-Based Approach. 7th Hellenic Conference on Informatics. August 27-29. Ioannina, Greece 1999.
#' @keywords data
NULL


R/rucrdtw-package.r
#' rucrdtw: Fast time series subsequence search in R
#'
#' @description Dynamic Time Warping (DTW) methods provide algorithms to optimally map a given time series onto all or part of another time series. The remaining cumulative distance between the series after the alignment is a useful distance metric in time series data mining applications for tasks such as classification, clustering, and anomaly detection. A broad suite of DTW algorithms is implemented in R in the \strong{dtw} package (Giorgino 2009).
#'
#' Calculating a DTW alignment is computationally relatively expensive, and as a consequence DTW is often a bottleneck in time series data mining applications. The UCR Suite (Rakthanmanon et al. 2012) provides a highly optimized algorithm for best-match subsequence searches that avoids unnecessary distance computations and thereby enables fast DTW and Euclidean Distance queries even in data sets containing trillions of observations.
#'
#' The \strong{rucrdtw} package provides R bindings for the UCR Suite. In addition to queries and data stored in text files, rucrdtw also implements methods for queries and/or data that are held in memory as R objects, as well as a method to do fast similarity searches against reference libraries of time series. The following table gives a quick overview over the different methods provided by \code{rucrdtw}:
#'
#' \tabular{lllll}{
#'   \strong{DTW method} \tab \strong{ED method} \tab \strong{data format} \tab \strong{query format} \tab \strong{Use case}\cr
#'   \code{\link{ucrdtw_ff}} \tab \code{\link{ucred_ff}} \tab text file \tab text file \tab data sets that are too large to keep in memory \cr
#'   \code{\link{ucrdtw_fv}} \tab \code{\link{ucred_fv}} \tab text file \tab  numeric vector \tab data sets that are too large to keep in memory\cr
#'   \code{\link{ucrdtw_vv}} \tab \code{\link{ucred_vv}} \tab numeric vector \tab numeric vector \tab data sets that fit into memory\cr
#'   \code{\link{ucrdtw_mv}} \tab \code{\link{ucred_mv}} \tab numeric matrix \tab numeric vector \tab compare query to a set of reference sequences of equal length
#' }
#'
#' Examples of the functionality in this package are provided in the rucrdtw vignette:
#'
#' \code{vignette("using_rucrdtw", package = "rucrdtw")}

#' @references Rakthanmanon, Thanawin, Bilson Campana, Abdullah Mueen, Gustavo Batista, Brandon Westover, Qiang Zhu, Jesin Zakaria, and Eamonn Keogh. 2012. Searching and Mining Trillions of Time Series Subsequences Under Dynamic Time Warping. In Proceedings of the 18th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, 262-70. ACM. doi:\doi{10.1145/2339530.2339576}.
#' @references Giorgino, Toni (2009). Computing and Visualizing Dynamic Time Warping Alignments in R: The dtw Package. Journal of Statistical Software, 31(7), 1-24, \doi{10.18637/jss.v031.i07}.
#' @references UCR Suite Website: \url{http://www.cs.ucr.edu/~eamonn/UCRsuite.html}
#' @seealso \link[dtw]{dtw-package}
#' @keywords internal
"_PACKAGE"

NULL


R/RcppExports.R
# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' UCR DTW Algorithm file-file method
#'
#' Sliding-window similarity search using DTW distance. This implementation is very close to the UCR Suite command line utility, in that it takes files as inputs for both query and data
#'
#' @name ucrdtw_ff
#' @param data character; path to data file
#' @param query character; path to query file
#' @param qlength integer; length of the query (number of data points), usually this should be equivalent to length(query), but it can be shorter.
#' @param dtwwindow double; Size of the warping window size (as a proportion of query length). The DTW calculation in `rucrdtw` uses a symmetric Sakoe-Chiba band. See Giorgino (2009) for a general coverage of warping window constraints.
#' @return a ucrdtw object. A list with the following elements
#' \itemize{
#'   \item \strong{location:} The starting location of the nearest neighbor of the given query, of size \code{length(query)}, in the data. Note that location starts from 1.
#'   \item \strong{distance:} The DTW distance between the nearest neighbor and the query.
#'   \item \strong{prunedKim:} Percentage of subsequences that were pruned based on the LB-Kim criterion.
#'   \item \strong{prunedKeogh:} Percentage of subsequences that were pruned based on the LB-Keogh-EQ criterion.
#'   \item \strong{prunedKeogh2:} Percentage of subsequences that were pruned based on the LB-Keogh-EC criterion.
#'   \item \strong{dtwCalc:} Percentage of subsequences for which the full DTW distance was calculated.
#' }
#' For an explanation of the pruning criteria see Rakthanmanon et al. (2012).
#' @references Rakthanmanon, Thanawin, Bilson Campana, Abdullah Mueen, Gustavo Batista, Brandon Westover, Qiang Zhu, Jesin Zakaria, and Eamonn Keogh. 2012. Searching and Mining Trillions of Time Series Subsequences Under Dynamic Time Warping. In Proceedings of the 18th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, 262-70. ACM. doi:\doi{10.1145/2339530.2339576}.
#' @references Giorgino, Toni (2009). Computing and Visualizing Dynamic Time Warping Alignments in R: The dtw Package. Journal of Statistical Software, 31(7), 1-24, doi:\doi{10.18637/jss.v031.i07}.
#' @examples
#' #locate example data file
#' dataf <- system.file("extdata/col_sc.txt", package="rucrdtw")
#' #locate example query file
#' queryf <- system.file("extdata/mid_sc.txt", package="rucrdtw")
#' #determine length of query file
#' qlength <- length(scan(queryf))
#' #run query
#' ucrdtw_ff(dataf, queryf, qlength, 0.05)
#' @useDynLib rucrdtw, .registration=TRUE
#' @importFrom Rcpp sourceCpp
#' @export
ucrdtw_ff <- function(data, query, qlength, dtwwindow) {
    .Call(`_rucrdtw_ucrdtw_ff`, data, query, qlength, dtwwindow)
}

#' UCR DTW Algorithm file-vector method
#'
#' Sliding-window similarity search using DTW distance. This implementation of the UCR Suite command line utility, takes a data file as input and an R numeric vector for the query
#'
#' @name ucrdtw_fv
#' @param data character; path to data file
#' @param query numeric vector containing the query. The query length is set to the length of this object.
#' @param dtwwindow double; Size of the warping window size (as a proportion of query length). The DTW calculation in `rucrdtw` uses a symmetric Sakoe-Chiba band. See Giorgino (2009) for a general coverage of warping window constraints.
#' @return a ucrdtw object. A list with the following elements
#' \itemize{
#'   \item \strong{location:} The starting location of the nearest neighbor of the given query, of size \code{length(query)}, in the data. Note that location starts from 1.
#'   \item \strong{distance:} The DTW distance between the nearest neighbor and the query.
#'   \item \strong{prunedKim:} Percentage of subsequences that were pruned based on the LB-Kim criterion.
#'   \item \strong{prunedKeogh:} Percentage of subsequences that were pruned based on the LB-Keogh-EQ criterion.
#'   \item \strong{prunedKeogh2:} Percentage of subsequences that were pruned based on the LB-Keogh-EC criterion.
#'   \item \strong{dtwCalc:} Percentage of subsequences for which the full DTW distance was calculated.
#' }
#' For an explanation of the pruning criteria see Rakthanmanon et al. (2012).
#' @references Rakthanmanon, Thanawin, Bilson Campana, Abdullah Mueen, Gustavo Batista, Brandon Westover, Qiang Zhu, Jesin Zakaria, and Eamonn Keogh. 2012. Searching and Mining Trillions of Time Series Subsequences Under Dynamic Time Warping. In Proceedings of the 18th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, 262-70. ACM. doi:\doi{10.1145/2339530.2339576}.
#' @references Giorgino, Toni (2009). Computing and Visualizing Dynamic Time Warping Alignments in R: The dtw Package. Journal of Statistical Software, 31(7), 1-24, doi:\doi{10.18637/jss.v031.i07}.
#' @examples
#' #locate example data file
#' dataf <- system.file("extdata/col_sc.txt", package="rucrdtw")
#' #load example data set
#' data("synthetic_control")
#' #extract first row as query
#' query <- synthetic_control[1,]
#' #run query
#' ucrdtw_fv(dataf, query, 0.05)
#' @useDynLib rucrdtw, .registration=TRUE
#' @importFrom Rcpp sourceCpp
#' @export
ucrdtw_fv <- function(data, query, dtwwindow) {
    .Call(`_rucrdtw_ucrdtw_fv`, data, query, dtwwindow)
}

#' UCR DTW Algorithm vector-vector method
#'
#' Sliding-window similarity search using DTW distance. This implementation of the UCR Suite command line utility, takes an R numeric vector as data input and an R numeric vector for the query
#'
#' @name ucrdtw_vv
#' @param data numeric vector containing data
#' @param query numeric vector containing the query. The length of this vector determines the query length.
#' @param dtwwindow double; Size of the warping window size (as a proportion of query length). The DTW calculation in `rucrdtw` uses a symmetric Sakoe-Chiba band. See Giorgino (2009) for a general coverage of warping window constraints.
#' @param epoch int defaults to 1e5, should be \eqn{\le} 1e6. This is the size of the data chunk that is processed at once. All cumulative values in the algorithm will be restarted after \code{epoch} iterations to reduce floating point errors in these values.
#' @param skip bool defaults to FALSE. If TRUE bound calculations and if necessary, distance calculations, are only performed on non-overlapping segments of the data (i.e. multiples of \code{qlength}). This is useful if \code{data} is a set of multiple reference time series, each of length \code{qlength}. The location returned when skipping is the index of the subsequence.
#' @return a ucrdtw object. A list with the following elements
#' \itemize{
#'   \item \strong{location:} The starting location of the nearest neighbor of the given query, of size \code{length(query)}, in the data. Note that location starts from 1.
#'   \item \strong{distance:} The DTW distance between the nearest neighbor and the query.
#'   \item \strong{prunedKim:} Percentage of subsequences that were pruned based on the LB-Kim criterion.
#'   \item \strong{prunedKeogh:} Percentage of subsequences that were pruned based on the LB-Keogh-EQ criterion.
#'   \item \strong{prunedKeogh2:} Percentage of subsequences that were pruned based on the LB-Keogh-EC criterion.
#'   \item \strong{dtwCalc:} Percentage of subsequences for which the full DTW distance was calculated.
#' }
#' For an explanation of the pruning criteria see Rakthanmanon et al. (2012).
#' @references Rakthanmanon, Thanawin, Bilson Campana, Abdullah Mueen, Gustavo Batista, Brandon Westover, Qiang Zhu, Jesin Zakaria, and Eamonn Keogh. 2012. Searching and Mining Trillions of Time Series Subsequences Under Dynamic Time Warping. In Proceedings of the 18th ACM SIGKDD International Conference on Knowledge Discovery and Data Mining, 262-70. ACM. doi:\doi{10.1145/2339530.2339576}.
#' @references Giorgino, Toni (2009). Computing and Visualizing Dynamic Time Warping Alignments in R: The dtw Package. Journal of Statistical Software, 31(7), 1-24, doi:\doi{10.18637/jss.v031.i07}.
#' @examples
#' #read example data into vector
#' datav <- scan(system.file("extdata/col_sc.txt", package="rucrdtw"))
#' #read example query into vector
#' query <- scan(system.file("extdata/first_sc.txt", package="rucrdtw"))
#' #execute query
#' ucrdtw_vv(datav, query, 0.05)
#' @useDynLib rucrdtw, .registration=TRUE
#' @importFrom Rcpp sourceCpp
#' @export
ucrdtw_vv <- function(data, query, dtwwindow, epoch = 100000L, skip = FALSE) {
    .Call(`_rucrdtw_ucrdtw_vv`, data, query, dtwwindow, epoch, skip)
}

#' UCR ED Algorithm file-file method
#'
#' Sliding-window similarity search using euclidean distance. This implementation is very close to the UCR Suite command line utility, in that it takes files as inputs for both query and data
#'
#' @name ucred_ff
#' @param data character; path to data file
#' @param query character; path to query file
#' @param qlength integer; length of query (n data points). Usually the length of the data contained in \code{query}, but it can be shorter.
#' @return a ucred object. A list with the following elements
#' \itemize{
#'   \item \strong{location:} The starting location of the nearest neighbor of the given query, of size \code{qlength}, in the data. Note that location starts from 1.
#'   \item \strong{distance:} The euclidean distance between the nearest neighbor and the query.
#' }
#' @examples
#' #locate example data file
#' dataf <- system.file("extdata/col_sc.txt", package="rucrdtw")
#' #locate example query file
#' queryf <- system.file("extdata/mid_sc.txt", package="rucrdtw")
#' #determine length of query file
#' qlength <- length(scan(queryf))
#' #run query
#' ucred_ff(dataf, queryf, qlength)
#' @useDynLib rucrdtw, .registration=TRUE
#' @importFrom Rcpp sourceCpp
#' @export
ucred_ff <- function(data, query, qlength) {
    .Call(`_rucrdtw_ucred_ff`, data, query, qlength)
}

#' UCR ED Algorithm file-vector method
#'
#' Sliding-window similarity search using Euclidean Distance. This implementation of the UCR Suite command line utility, takes a data file as input and an R numeric vector for the query.
#'
#' @name ucred_fv
#' @param data character; path to data file
#' @param query numeric vector containing query
#' @return a ucred object. A list with the following elements
#' \itemize{
#'   \item \strong{location:} The starting location of the nearest neighbor of the given query, of size \code{length(query)}, in the data. Note that location starts from 1.
#'   \item \strong{distance:} The Euclidean Distance between the nearest neighbor and the query.
#' }
#' @examples
#' #locate example data file
#' dataf <- system.file("extdata/col_sc.txt", package="rucrdtw")
#' #read example query file into vector
#' query <- scan(system.file("extdata/mid_sc.txt", package="rucrdtw"))
#' #run query
#' ucred_fv(dataf, query)
#' @useDynLib rucrdtw, .registration=TRUE
#' @importFrom Rcpp sourceCpp
#' @export
ucred_fv <- function(data, query) {
    .Call(`_rucrdtw_ucred_fv`, data, query)
}

#' UCR ED Algorithm vector-vector method
#'
#' Sliding-window similarity search using Euclidean Distance. This implementation of the UCR Suite Euclidean Distance command line utility takes an R numeric vector as data input and an R numeric vector for the query.
#'
#' @name ucred_vv
#' @param data numeric vector containing data
#' @param query numeric vector containing query
#' @param skip bool defaults to TRUE. If TRUE bound calculations and if necessary, distance calculations, are only performed on non-overlapping segments of the data (i.e. multiples of \code{length(query)}). This is useful if \code{data} is a set of multiple reference time series, each of length \code{length(query)}. The location returned when skipping is the index of the subsequence.
#' @return a ucred object. A list with the following elements
#' \itemize{
#'   \item \strong{location:} The starting location of the nearest neighbor of the given query, of size \code{length(query)}, in the data. Note that location starts from 1.
#'   \item \strong{distance:} The Euclidean Distance between the nearest neighbor and the query.
#' }
#' @examples
#' #read example file into vector
#' dataf <- scan(system.file("extdata/col_sc.txt", package="rucrdtw"))
#' #read example query file into vector
#' query <- scan(system.file("extdata/mid_sc.txt", package="rucrdtw"))
#' #run query
#' ucred_vv(dataf, query)
#' @useDynLib rucrdtw, .registration=TRUE
#' @importFrom Rcpp sourceCpp
#' @export
ucred_vv <- function(data, query, skip = FALSE) {
    .Call(`_rucrdtw_ucred_vv`, data, query, skip)
}

# Register entry points for exported C++ functions
methods::setLoadAction(function(ns) {
    .Call(`_rucrdtw_RcppExport_registerCCallable`)
})
